### ISSUE TITLE
compile error with serialization

### ISSUE THREAD

**Issue by radarsat1** [siconos/siconos#328]
Here are the complete flags I am trying to compile with (debian unstable):

```
$ cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_INSTALL_RUNSTATEDIR=/run "-GUnix Makefiles" -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_AUTOGEN_VERBOSE=ON -DCMAKE_INSTALL_LIBDIR=lib/x86_64-linux-gnu -DCMAKE_SKIP_BUILD_RPATH=TRUE -DFORCE_SKIP_RPATH=1 -DCMAKE_SKIP_INSTALL_RPATH=TRUE -DCMAKE_SKIP_RPATH=TRUE -DINSTALL_CMAKE_DIR=/usr/lib/x86_64-linux-gnu/cmake/siconos -DPYTHON_EXECUTABLE=/usr/bin/python3 -DI_WANT_STATIC_LPSOLVE=ON -DLpSolve_LIBRARY=/usr/lib/liblpsolve55_pic.a -DWITH_TESTING=ON -DWITH_DOCUMENTATION=ON -DWITH_DOXY2SWIG=ON -DWITH_SYSTEM_SUITESPARSE=ON -DWITH_SERIALIZATION=ON -DWITH_BULLET=ON -DWITH_FCLIB=OFF -DWITH_OCE=OFF -DIDONTWANTMPI=ON -DWITH_MUMPS=ON ..
```

The compilation ends with,
```
/home/sinclairs/siconos/bld/wrap/siconos/numericsPYTHON_wrap.cxx:3570:10: fatal error: SiconosFullNumerics.hpp: No such file or directory
 3570 | #include <SiconosFullNumerics.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Comment by radarsat1** [comment]
The path io/numerics/src is _not_ in the gcc command, not sure why:

```
cd /home/sinclairs/siconos/bld/numerics/swig && /usr/bin/c++  -D_numerics_EXPORTS -I/home/sinclairs/siconos/bld -I/home/sinclairs/siconos/wrap/swig -I/home/sinclairs/siconos/bld/wrap/siconos -I/home/sinclairs/siconos/numerics/swig -I/home/sinclairs/siconos/io/src -I/home/sinclairs/siconos/io/src/mechanics -I/home/sinclairs/siconos/numerics/src -I/home/sinclairs/siconos/numerics/src/tools -I/home/sinclairs/siconos/numerics/src/tools/InterfaceToPathFerris -I/home/sinclairs/siconos/numerics/src/tools/internal -I/home/sinclairs/siconos/numerics/src/AVI -I/home/sinclairs/siconos/numerics/src/LCP -I/home/sinclairs/siconos/numerics/src/MLCP -I/home/sinclairs/siconos/numerics/src/SOCP -I/home/sinclairs/siconos/numerics/src/QP -I/home/sinclairs/siconos/numerics/src/Relay -I/home/sinclairs/siconos/numerics/src/FrictionContact -I/home/sinclairs/siconos/numerics/src/FrictionContact/Generated -I/home/sinclairs/siconos/numerics/src/NCP -I/home/sinclairs/siconos/numerics/src/MCP -I/home/sinclairs/siconos/numerics/src/GenericMechanical -I/home/sinclairs/siconos/numerics/src/VI -I/home/sinclairs/siconos/externals/netlib -I/home/sinclairs/siconos/externals/netlib/odepack -I/home/sinclairs/siconos/externals/tools -I/home/sinclairs/siconos/externals/blas_lapack -I/home/sinclairs/siconos/externals/SOL/lumod-c -I/home/sinclairs/siconos/externals/hairer -I/home/sinclairs/siconos/externals/numeric_bindings -I/home/sinclairs/siconos/externals/boost_contribs -I/home/sinclairs/siconos/externals/sort -isystem /usr/include/python3.8 -isystem /usr/include/suitesparse -isystem /usr/lib/python3/dist-packages/numpy/core/include  -fPIC   -std=gnu++11 -fno-strict-aliasing -o CMakeFiles/_numerics.dir/__/__/wrap/siconos/numericsPYTHON_wrap.cxx.o -c /home/sinclairs/siconos/bld/wrap/siconos/numericsPYTHON_wrap.cxx
```

**Comment by radarsat1** [comment]
Sorry I meant, the path to `io/src/serialization/` which contains SiconosFullNumerics.hpp

**Comment by radarsat1** [comment]
I can confirm that I get the same result with `-DUSER_OPTIONS_FILE`,

> ci_gitlab/siconos_confs/siconos_with_serialization.cmake

**Comment by radarsat1** [comment]
Okay *this* problem is resolved by,

```
diff --git a/numerics/swig/CMakeLists.txt b/numerics/swig/CMakeLists.txt
index d1fc01dc9..bae93bcfd 100644
--- a/numerics/swig/CMakeLists.txt
+++ b/numerics/swig/CMakeLists.txt
@@ -53,14 +53,16 @@ swig_module_setup(${COMPONENT}_PYTHON_MODULES)
 
 target_link_libraries(_numerics $<BUILD_INTERFACE:externals>)
 
+if(WITH_SERIALIZATION)
+  include_directories(${CMAKE_SOURCE_DIR}/io/src/serialization)
+endif()
+
 # If WITH_SYSTEM_SUITESPARSE, suite sparse is an imported target
 # that must be taken into account by swig process.
 if(WITH_SYSTEM_SUITESPARSE)
   target_link_libraries(_numerics $<BUILD_INTERFACE:SuiteSparse::CXSparse>)
 endif()
```

to `numerics/swig/CMakeLists.txt` at line 56.

But this leads to an error with SolverOptions which is resolved by,

```
diff --git a/io/src/serialization/SiconosFullNumerics.hpp b/io/src/serialization/SiconosFullNumerics.hpp
index 1c923f53b..78a924b32 100644
--- a/io/src/serialization/SiconosFullNumerics.hpp
+++ b/io/src/serialization/SiconosFullNumerics.hpp
@@ -40,7 +40,7 @@ void siconos_io(Archive& ar, SolverOptions&v, unsigned int version)
   {
     v.iparam = (int *) malloc(v.iSize * sizeof(int));
     v.dparam = (double *) malloc(v.dSize * sizeof(double));
-    v.internalSolvers = (SolverOptions *) malloc(v.numberOfInternalSolvers * sizeof(SolverOptions));
+    v.internalSolvers = (SolverOptions**)calloc(v.numberOfInternalSolvers, sizeof(SolverOptions*));
     v.callback = (Callback *) malloc(sizeof(Callback));
   }
   SERIALIZE(v, (callback), ar);
```

Unfortunately then the build dies on generation:

```
[ 82%] Generating SiconosFullGenerated.hpp
In file included from /home/sinclairs/siconos/kernel/src/utils/SiconosAlgebra/SiconosMatrix.hpp:29,
                 from /home/sinclairs/siconos/kernel/src/utils/SiconosAlgebra/SimpleMatrix.hpp:30,
                 from /home/sinclairs/siconos/kernel/src/modelingTools/DynamicalSystem.hpp:33,
                 from /home/sinclairs/siconos/kernel/src/simulationTools/InteractionManager.hpp:27,
                 from /home/sinclairs/siconos/mechanics/src/collision/native/SpaceFilter.hpp:42,
                 from /tmp/tmpgk8ot424/headers.hpp:2:
/home/sinclairs/siconos/numerics/src/tools/CSparseMatrix.h:49:10: fatal error: cs.h: No such file or directory
   49 | #include "cs.h"
      |          ^~~~~~
compilation terminated.
```

**Comment by radarsat1** [comment]
Rest of the output:

```
Traceback (most recent call last):
  File "../../io/tools/build_from_doxygen.py", line 153, in <module>
    header_classes = classes_from_headers(all_headers, include_paths)
  File "../../io/tools/build_from_doxygen.py", line 131, in classes_from_headers
    with open(compiled, 'r') as out:
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpgk8ot424/out.cpp'
make[2]: *** [io/CMakeFiles/SerializersGeneration.dir/build.make:66: io/SiconosFullGenerated.hpp] Error 1
make[1]: *** [CMakeFiles/Makefile2:5861: io/CMakeFiles/SerializersGeneration.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

**Comment by radarsat1** [comment]
Okay, this appears solved with,

``
diff --git a/io/CMakeLists.txt b/io/CMakeLists.txt
index 97221c277..d78be2291 100644
--- a/io/CMakeLists.txt
+++ b/io/CMakeLists.txt
@@ -63,6 +63,9 @@ if(WITH_SERIALIZATION)
       list(APPEND GENERATED_DEPENDENCIES ${CMAKE_SOURCE_DIR}/control/src/SiconosControl.hp
p)
     endif()
     list(APPEND GENERATED_INCLUDES -I${CMAKE_BINARY_DIR}) # For SiconosConfig.h
+    if(SuiteSparse_CXSparse_INCLUDE_DIR)
+      list(APPEND GENERATED_INCLUDES -I${SuiteSparse_CXSparse_INCLUDE_DIR})
+    endif()
     
     if(MPI_FOUND)
       foreach(_DIR IN LISTS MPI_CXX_INCLUDE_PATH)
```

But now,

```
 [ 82%] Generating SiconosFullGenerated.hpp
0 classes found.
../../io/tools/build_from_doxygen.py: Error, not enough classes found.
make[2]: *** [io/CMakeFiles/SerializersGeneration.dir/build.make:66: io/SiconosFullGenerated.hpp] Error 1
make[1]: *** [CMakeFiles/Makefile2:5861: io/CMakeFiles/SerializersGeneration.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

So generation (and therefore, it seems, serialization) is still broken.  I guess the build was changed to always do SerializersGeneration, sorry I wasn't following that.  There must be a problem scanning the headers in the build tool, hopefully I can look into it soon otherwise someone please take a look.

**Comment by radarsat1** [comment]
I added a branch to my fork with these changes, as it seems i cannot push to the main siconos repository. https://github.com/radarsat1/siconos/pull/new/fix-serializersgeneration

**Comment by vacary** [comment]
Now, we push only on the gricad side. There is a mirroring

**Comment by radarsat1** [comment]
Ok thanks! but I didn't meant to merge this to master yet ;)  But no problem, hopefully they are good changes, still some things to work out.

**Comment by radarsat1** [comment]
So a problem blocking serialization is that the classes for generated serialization are taken from doxygen's XML output, and it seems that doxygen is silently not documenting all classes!  For example, there is no doc for BlockCSRMatrix.  I discovered that if I remove the `\rst` line in the doc comment for that class, then it does get commented!  doxygen bug???  what to do?

Here is a little interaction demonstrating it... before removing `\rst`, only docs for the header file are generated, but not for the class.  I remove `\rst` using sed and then it generates docs for the class:

```
sinclairs@035d759ffae4:~/siconos/build (master)
$ find . -name \*BlockCSR\*
./docs/build/html/doxygen/BlockCSRMatrix_8hpp.html
./docs/build/html/doxygen/BlockCSRMatrix_8hpp_source.html

sinclairs@035d759ffae4:~/siconos/build (master)
$ doxygen config/doxy.config
warning: Tag 'PERL_PATH' at line 2125 of file 'config/doxy.config' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'MSCGEN_PATH' at line 2147 of file 'config/doxy.config' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"

sinclairs@035d759ffae4:~/siconos/build (master)
$ find . -name \*BlockCSR\*
./docs/build/html/doxygen/BlockCSRMatrix_8hpp.html
./docs/build/html/doxygen/BlockCSRMatrix_8hpp_source.html

sinclairs@035d759ffae4:~/siconos/build (master)
$ sed -ie 's,\\rst,,' ../kernel/src/simulationTools/BlockCSRMatrix.hpp

sinclairs@035d759ffae4:~/siconos/build (master)
$ doxygen config/doxy.config
warning: Tag 'PERL_PATH' at line 2125 of file 'config/doxy.config' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'MSCGEN_PATH' at line 2147 of file 'config/doxy.config' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"

sinclairs@035d759ffae4:~/siconos/build (master)
$ find . -name \*BlockCSR\*
./docs/build/html/doxygen/classBlockCSRMatrix.html
./docs/build/html/doxygen/BlockCSRMatrix_8hpp.html
./docs/build/html/doxygen/classBlockCSRMatrix-members.html
./docs/build/html/doxygen/classBlockCSRMatrix__coll__graph.map
./docs/build/html/doxygen/classBlockCSRMatrix__coll__graph.md5
./docs/build/html/doxygen/classBlockCSRMatrix__coll__graph.png
./docs/build/html/doxygen/BlockCSRMatrix_8hpp_source.html
```

**Comment by radarsat1** [comment]
```
$ doxygen --version
1.8.16
```

It would be good to verify if this did _not_ happen on previous versions of doxygen.

**Comment by radarsat1** [comment]
Turning on doxygen warnings, finally finding some sources of "silent" problems that are causing classes to go undocumented, eg.,

```
$ doxygen docs/config/kerneldoxy-xml.config 2>&1 | grep FirstOrderR
/home/sinclairs/siconos/kernel/src/modelingTools/FirstOrderR.hpp:179: warning: Reached end of file while still inside a (nested) comment. Nesting level 1 (probable line reference: 29)
/home/sinclairs/siconos/kernel/src/modelingTools/FirstOrderR.hpp:179: warning: reached end of file while inside a 'f$' block!
```

**Comment by radarsat1** [comment]
Most problems have been addressed save for failure when bullet is enabled.
A patch that makes it compile can be found here, but obviously this disabled important stuff so it shoul not be merged, just posting it for informational purposes: https://github.com/radarsat1/siconos/commit/7413d667cfe390690560b5bf3a68d6d996605e56

**Comment by radarsat1** [comment]
This is all resolved now as of 39bfdd6a161a467e86fe0a05bc6031cba80f4c52
